File: backend/app/schemas/finance.py
"""Pydantic schemas for Unified Billing, Invoicing, Payments, Financial Operations & Commercial Control."""

from datetime import datetime
from decimal import Decimal
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field


# --- Billing Profile ---

class BillingProfileCreate(BaseModel):
    client_id: str
    legal_name: str
    tax_id: Optional[str] = None
    billing_email: str
    address_line1: Optional[str] = None
    address_line2: Optional[str] = None
    city: Optional[str] = None
    state: Optional[str] = None
    postal_code: Optional[str] = None
    country: str = "US"
    default_currency: str = "USD"
    payment_terms: str = "net_30"


class BillingProfileResponse(BaseModel):
    id: str
    tenant_id: str
    client_id: str
    legal_name: str
    tax_id: Optional[str] = None
    billing_email: str
    address_line1: Optional[str] = None
    address_line2: Optional[str] = None
    city: Optional[str] = None
    state: Optional[str] = None
    postal_code: Optional[str] = None
    country: str
    default_currency: str
    payment_terms: str
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# --- Currency & Tax Profiles ---

class CurrencyCreate(BaseModel):
    code: str
    name: str
    symbol: str = "$"
    decimal_places: int = 2
    is_active: bool = True


class CurrencyResponse(BaseModel):
    id: str
    code: str
    name: str
    symbol: str
    decimal_places: int
    is_active: bool
    created_at: datetime

    class Config:
        from_attributes = True


class TaxProfileCreate(BaseModel):
    name: str
    tax_type: str = "sales_tax"
    rate_pct: Decimal = Decimal("0.00")
    country: str
    state: Optional[str] = None
    is_compound: bool = False
    is_active: bool = True


class TaxProfileResponse(BaseModel):
    id: str
    tenant_id: str
    name: str
    tax_type: str
    rate_pct: Decimal
    country: str
    state: Optional[str] = None
    is_compound: bool
    is_active: bool
    created_at: datetime

    class Config:
        from_attributes = True


# --- Line Items & Invoices ---

class InvoiceItemCreate(BaseModel):
    title: str
    description: Optional[str] = None
    quantity: Decimal = Decimal("1.00")
    unit_price: Decimal = Decimal("0.00")
    item_type: str = "fixed"
    reference_type: Optional[str] = None
    reference_id: Optional[str] = None
    contract_id: Optional[str] = None


class InvoiceItemResponse(BaseModel):
    id: str
    invoice_id: str
    title: str
    description: Optional[str] = None
    quantity: Decimal
    unit_price: Decimal
    subtotal: Decimal
    discount_amount: Decimal
    taxable_amount: Decimal
    tax_amount: Decimal
    total_amount: Decimal
    item_type: str
    reference_type: Optional[str] = None
    reference_id: Optional[str] = None
    contract_id: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True


class InvoiceCreate(BaseModel):
    client_id: str
    billing_profile_id: Optional[str] = None
    project_id: Optional[str] = None
    contract_id: Optional[str] = None
    proposal_id: Optional[str] = None
    currency: str = "USD"
    payment_terms: str = "net_30"
    custom_terms_days: Optional[int] = None
    items: List[InvoiceItemCreate] = []
    discount_rate_pct: Decimal = Decimal("0.00")
    fixed_discount_amount: Decimal = Decimal("0.00")
    tax_rate_pct: Decimal = Decimal("0.00")
    tax_type: str = "none"
    tax_region: Optional[str] = None
    notes: Optional[str] = None
    terms_and_conditions: Optional[str] = None


class InvoiceUpdate(BaseModel):
    billing_profile_id: Optional[str] = None
    payment_terms: Optional[str] = None
    custom_terms_days: Optional[int] = None
    items: Optional[List[InvoiceItemCreate]] = None
    discount_rate_pct: Optional[Decimal] = None
    fixed_discount_amount: Optional[Decimal] = None
    tax_rate_pct: Optional[Decimal] = None
    tax_type: Optional[str] = None
    tax_region: Optional[str] = None
    notes: Optional[str] = None
    terms_and_conditions: Optional[str] = None


class InvoiceResponse(BaseModel):
    id: str
    tenant_id: str
    client_id: str
    billing_profile_id: Optional[str] = None
    project_id: Optional[str] = None
    contract_id: Optional[str] = None
    proposal_id: Optional[str] = None
    invoice_number: str
    version: int
    status: str
    currency: str
    payment_terms: str
    custom_terms_days: Optional[int] = None
    issue_date: Optional[datetime] = None
    due_date: Optional[datetime] = None
    subtotal: Decimal
    discount_amount: Decimal
    taxable_amount: Decimal
    tax_amount: Decimal
    total_amount: Decimal
    amount_paid: Decimal
    balance_due: Decimal
    tax_rate_pct: Decimal
    tax_type: str
    tax_region: Optional[str] = None
    discount_rate_pct: Decimal
    fixed_discount_amount: Decimal
    notes: Optional[str] = None
    terms_and_conditions: Optional[str] = None
    created_by_user_id: Optional[str] = None
    issued_by_user_id: Optional[str] = None
    approved_by_user_id: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class InvoiceDetailResponse(InvoiceResponse):
    items: List[InvoiceItemResponse] = []


class InvoiceApprovalRequest(BaseModel):
    status: str  # approved, rejected
    comments: Optional[str] = None


# --- Payments & Refunds ---

class PaymentChargeRequest(BaseModel):
    invoice_id: str
    amount: Decimal
    currency: str = "USD"
    payment_method: str = "credit_card"
    provider_type: Optional[str] = "mock"
    customer_id: Optional[str] = None
    idempotency_key: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None


class PaymentResponse(BaseModel):
    id: str
    tenant_id: str
    invoice_id: Optional[str] = None
    client_id: str
    provider_type: str
    provider_payment_id: Optional[str] = None
    amount: Decimal
    currency: str
    status: str
    payment_method: str
    fee_amount: Decimal
    error_code: Optional[str] = None
    error_message: Optional[str] = None
    created_by_user_id: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True


class RefundProcessRequest(BaseModel):
    payment_id: str
    amount: Decimal
    currency: str = "USD"
    reason: str = "Customer requested refund"
    idempotency_key: Optional[str] = None


class RefundResponse(BaseModel):
    id: str
    tenant_id: str
    payment_id: str
    invoice_id: Optional[str] = None
    provider_refund_id: Optional[str] = None
    amount: Decimal
    currency: str
    status: str
    reason: Optional[str] = None
    error_code: Optional[str] = None
    error_message: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True


# --- Reconciliation ---

class PaymentReconciliationResponse(BaseModel):
    id: str
    tenant_id: str
    payment_id: Optional[str] = None
    provider_event_id: Optional[str] = None
    invoice_id: Optional[str] = None
    status: str
    matched_amount: Decimal
    discrepancy_amount: Decimal
    discrepancy_type: Optional[str] = None
    notes: Optional[str] = None
    resolved_by_user_id: Optional[str] = None
    resolved_at: Optional[datetime] = None
    created_at: datetime

    class Config:
        from_attributes = True


# --- Credit Note ---

class CreditNoteCreate(BaseModel):
    client_id: str
    invoice_id: Optional[str] = None
    amount: Decimal
    currency: str = "USD"
    reason: Optional[str] = None


class CreditNoteResponse(BaseModel):
    id: str
    tenant_id: str
    client_id: str
    invoice_id: Optional[str] = None
    credit_note_number: str
    amount: Decimal
    currency: str
    balance: Decimal
    reason: Optional[str] = None
    status: str
    created_at: datetime

    class Config:
        from_attributes = True


# --- Financial Accounts & Ledger ---

class FinancialAccountCreate(BaseModel):
    account_code: str
    account_name: str
    account_type: str
    currency: str = "USD"
    is_active: bool = True


class FinancialAccountResponse(BaseModel):
    id: str
    tenant_id: str
    account_code: str
    account_name: str
    account_type: str
    currency: str
    is_active: bool
    created_at: datetime

    class Config:
        from_attributes = True


class JournalItem(BaseModel):
    account_id: str
    account_type: Optional[str] = None
    direction: str  # debit, credit
    amount: Decimal
    description: Optional[str] = None


class JournalEntryRequest(BaseModel):
    description: str
    reference_type: str
    reference_id: str
    currency: str = "USD"
    items: List[JournalItem]


class LedgerEntryResponse(BaseModel):
    id: str
    tenant_id: str
    entry_group_id: str
    account_id: str
    account_type: Optional[str] = None
    direction: str
    amount: Decimal
    currency: str
    reference_type: Optional[str] = None
    reference_id: Optional[str] = None
    description: Optional[str] = None
    is_reversed: bool
    posted_at: datetime
    created_at: datetime

    class Config:
        from_attributes = True


# --- Profitability & Forecast ---

class ProjectProfitabilityRequest(BaseModel):
    project_id: str
    contracted_revenue: Decimal
    invoiced_revenue: Decimal
    collected_revenue: Decimal
    estimated_cost: Decimal
    labor_hours_logged: Decimal = Decimal("0.00")
    labor_hourly_cost_rate: Decimal = Decimal("0.00")
    ai_token_cost: Decimal = Decimal("0.00")
    infrastructure_cost: Decimal = Decimal("0.00")
    subcontractor_cost: Decimal = Decimal("0.00")
    other_expenses: Decimal = Decimal("0.00")


class CashFlowForecastRequest(BaseModel):
    months_ahead: int = 3
    projected_monthly_burn: Decimal = Decimal("0.00")

File: frontend/lib/api/finance.ts
import { api } from './client';

export interface BillingProfile {
  id: string;
  tenant_id: string;
  client_id: string;
  legal_name: string;
  tax_id?: string;
  billing_email: string;
  address_line1?: string;
  address_line2?: string;
  city?: string;
  state?: string;
  postal_code?: string;
  country: string;
  default_currency: string;
  payment_terms: string;
  created_at: string;
  updated_at: string;
}

export interface TaxProfile {
  id: string;
  tenant_id: string;
  name: string;
  tax_type: string;
  rate_pct: string;
  country: string;
  state?: string;
  is_compound: boolean;
  is_active: boolean;
  created_at: string;
}

export interface Currency {
  id: string;
  code: string;
  name: string;
  symbol: string;
  decimal_places: number;
  is_active: boolean;
  created_at: string;
}

export interface InvoiceItem {
  id?: string;
  invoice_id?: string;
  title: string;
  description?: string;
  quantity: string;
  unit_price: string;
  subtotal?: string;
  discount_amount?: string;
  taxable_amount?: string;
  tax_amount?: string;
  total_amount?: string;
  item_type?: string;
  reference_type?: string;
  reference_id?: string;
  contract_id?: string;
}

export interface Invoice {
  id: string;
  tenant_id: string;
  client_id: string;
  billing_profile_id?: string;
  project_id?: string;
  contract_id?: string;
  proposal_id?: string;
  invoice_number: string;
  version: number;
  status: string;
  currency: string;
  payment_terms: string;
  custom_terms_days?: number;
  issue_date?: string;
  due_date?: string;
  subtotal: string;
  discount_amount: string;
  taxable_amount: string;
  tax_amount: string;
  total_amount: string;
  amount_paid: string;
  balance_due: string;
  tax_rate_pct: string;
  tax_type: string;
  tax_region?: string;
  discount_rate_pct: string;
  fixed_discount_amount: string;
  notes?: string;
  terms_and_conditions?: string;
  created_by_user_id?: string;
  issued_by_user_id?: string;
  approved_by_user_id?: string;
  created_at: string;
  updated_at: string;
  items?: InvoiceItem[];
}

export interface Payment {
  id: string;
  tenant_id: string;
  invoice_id?: string;
  client_id: string;
  provider_type: string;
  provider_payment_id?: string;
  amount: string;
  currency: string;
  status: string;
  payment_method: string;
  fee_amount: string;
  error_code?: string;
  error_message?: string;
  created_by_user_id?: string;
  created_at: string;
}

export interface ProjectProfitabilityAnalysis {
  project_id: string;
  financial_summary: {
    contracted_revenue: string;
    invoiced_revenue: string;
    collected_revenue: string;
    unbilled_contract_balance: string;
    outstanding_receivables: string;
  };
  cost_breakdown?: {
    estimated_cost: string;
    actual_total_cost: string;
    labor_hours_logged: string;
    direct_labor_cost: string;
    ai_token_cost: string;
    infrastructure_cost: string;
    subcontractor_cost: string;
    other_expenses: string;
    cost_variance_amount: string;
    cost_variance_pct: string;
  };
  profitability_metrics?: {
    estimated_gross_profit: string;
    estimated_margin_pct: string;
    invoiced_gross_profit: string;
    invoiced_margin_pct: string;
    realized_gross_profit: string;
    realized_margin_pct: string;
  };
  commercial_health?: {
    status: string;
    badge: string;
    is_cost_overrun: boolean;
    is_margin_compressed: boolean;
  };
  agent_insights?: string[];
  calculated_at: string;
}

export interface CashFlowForecast {
  months_ahead: number;
  projected_inflows: {
    outstanding_receivables: string;
    scheduled_billings: string;
    total_expected_inflow: string;
  };
  projected_outflows: {
    monthly_burn_rate: string;
    total_expected_burn: string;
  };
  net_projected_cash_flow: string;
  runway_indicator: string;
  calculated_at: string;
}

export const financeApi = {
  // Invoices
  listInvoices: (params?: { client_id?: string; project_id?: string; status?: string; limit?: number; offset?: number }) =>
    api.get<Invoice[]>('/invoices', { params }),

  getInvoice: (id: string) =>
    api.get<Invoice>(`/invoices/${id}`),

  createInvoice: (data: Partial<Invoice>) =>
    api.post<Invoice>('/invoices', data),

  approveInvoice: (id: string, status: 'approved' | 'rejected', comments?: string) =>
    api.post<Invoice>(`/invoices/${id}/approve`, { status, comments }),

  issueInvoice: (id: string, params?: { ar_account_id?: string; revenue_account_id?: string; tax_account_id?: string }) =>
    api.post<Invoice>(`/invoices/${id}/issue`, null, { params }),

  // Payments
  listPayments: (params?: { invoice_id?: string; client_id?: string }) =>
    api.get<Payment[]>('/payments', { params }),

  chargePayment: (data: {
    invoice_id: string;
    amount: string;
    currency?: string;
    payment_method?: string;
    provider_type?: string;
    customer_id?: string;
    idempotency_key?: string;
  }) =>
    api.post<Payment>('/payments/charge', data),

  processRefund: (data: { payment_id: string; amount: string; currency?: string; reason?: string }) =>
    api.post('/payments/refund', data),

  // Billing & Taxes
  listBillingProfiles: (clientId?: string) =>
    api.get<BillingProfile[]>('/billing/profiles', { params: { client_id: clientId } }),

  createBillingProfile: (data: Partial<BillingProfile>) =>
    api.post<BillingProfile>('/billing/profiles', data),

  listTaxProfiles: () =>
    api.get<TaxProfile[]>('/billing/taxes'),

  listCurrencies: () =>
    api.get<Currency[]>('/billing/currencies'),

  // Intelligence & Ledger
  analyzeProfitability: (data: {
    project_id: string;
    contracted_revenue: string;
    invoiced_revenue: string;
    collected_revenue: string;
    estimated_cost: string;
    labor_hours_logged?: string;
    labor_hourly_cost_rate?: string;
    ai_token_cost?: string;
    infrastructure_cost?: string;
    subcontractor_cost?: string;
    other_expenses?: string;
  }) =>
    api.post<ProjectProfitabilityAnalysis>('/finance/profitability/analyze', data),

  forecastCashFlow: (data: { months_ahead?: number; projected_monthly_burn?: string }) =>
    api.post<CashFlowForecast>('/finance/forecast/cash-flow', data),
};
